chiselFrontend/src/main/scala/chisel3/core/Aggregate.scala

```diff
@@ -320,50 +320,60 @@ trait VecLike[T <: Data] extends collection.IndexedSeq[T] with HasId {
     SeqUtils.oneHotMux(indexWhereHelper(p))
 }
 
-/** Base class for data types defined as a bundle of other data types.
+/** Base class for Aggregates based on key values pairs of String and Data
   *
-  * Usage: extend this class (either as an anonymous or named class) and define
-  * members variables of [[Data]] subtypes to be elements in the Bundle.
+  * Record should only be extended by libraries and fairly sophisticated generators.
+  * RTL writers should use [[Bundle]].
   */
-class Bundle extends Aggregate {
+abstract class Record extends Aggregate {
+
+  val elements: ListMap[String, Data]
+
+  /** Name for Pretty Printing */
+  def className: String = this.getClass.getSimpleName
+
   private val _namespace = Builder.globalNamespace.child
 
-  // TODO: replace with better defined FIRRTL weak-connect operator
-  /** Connect elements in this Bundle to elements in `that` on a best-effort
-    * (weak) basis, matching by type, orientation, and name.
-    *
-    * @note unconnected elements will NOT generate errors or warnings
-    *
-    * @example
-    * {{{
-    * // Pass through wires in this module's io to those mySubModule's io,
-    * // matching by type, orientation, and name, and ignoring extra wires.
-    * mySubModule.io <> io
-    * }}}
-    */
+  private[chisel3] def toType = {
+    def eltPort(elt: Data): String = {
+      val flipStr: String = if(Data.isFirrtlFlipped(elt)) "flip " else ""
+      s"${flipStr}${elt.getRef.name} : ${elt.toType}"
+    }
+    s"{${elements.toIndexedSeq.reverse.map(e => eltPort(e._2)).mkString(", ")}}"
+  }
 
-  lazy val elements: ListMap[String, Data] = ListMap(namedElts:_*)
+  private[chisel3] lazy val flatten = elements.toIndexedSeq.flatMap(_._2.flatten)
 
-  /** Returns a best guess at whether a field in this Bundle is a user-defined
-    * Bundle element without looking at type signatures.
-    */
-  private def isBundleField(m: java.lang.reflect.Method) =
-    m.getParameterTypes.isEmpty &&
-    !java.lang.reflect.Modifier.isStatic(m.getModifiers) &&
-    !(Bundle.keywords contains m.getName) && !(m.getName contains '$')
+  private[chisel3] override def _onModuleClose: Unit = // scalastyle:ignore method.name
+    for ((name, elt) <- elements) { elt.setRef(this, _namespace.name(name)) }
```

Comment:
PS: For future changes, _onModuleClose doesn't quite capture what the point of the function is (setting up dependent references). Particularly because this can (and probably should) be done before the Module is 'closed'.